http-clients/aws-crt-client/src/main/java/software/amazon/awssdk/http/crt/SocketOptionsConfiguration.java

```diff
@@ -0,0 +1,215 @@
+
+import software.amazon.awssdk.annotations.SdkPreviewApi;
+import software.amazon.awssdk.annotations.SdkPublicApi;
+import software.amazon.awssdk.crt.io.SocketOptions;
+import software.amazon.awssdk.utils.Validate;
+
+/**
+ * Configuration that defines socket options for all connections established by
+ * the {@link SocketOptionsConfiguration}.
+ *
+ * <b>NOTE:</b> This is a Preview API and is subject to change so it should not be used in production.
+ */
+@SdkPublicApi
+@SdkPreviewApi
+public final class SocketOptionsConfiguration {
+
+    private final SocketOptions.SocketDomain domain;
+    private final SocketOptions.SocketType type;
+    private final int connectTimeoutMs;
+    private final int keepAliveIntervalSecs;
+    private final int keepAliveTimeoutSecs;
+
+    private SocketOptionsConfiguration(DefaultSocketOptionsConfigurationBuilder builder) {
+        this.domain = Validate.paramNotNull(builder.domain,
+                                            "domain");
+        this.type = Validate.paramNotNull(builder.type,
+                                          "type");
+        this.connectTimeoutMs = Validate.isPositive(builder.connectTimeoutMs,
+                                                       "connectTimeoutMs");
+        this.keepAliveIntervalSecs = Validate.isNotNegative(builder.keepAliveIntervalSecs,
+                                                            "keepAliveIntervalSecs");
+        this.keepAliveTimeoutSecs = Validate.isNotNegative(builder.keepAliveTimeoutSecs,
+                                                           "keepAliveTimeoutSecs");
+    }
+
+    /**
+     * @return socket domain
+     */
+    public SocketOptions.SocketDomain domain() {
+        return domain;
+    }
+
+    /**
+     * @return socket type
+     */
+    public SocketOptions.SocketType type() {
+        return type;
+    }
+
+    /**
+     * @return number of milliseconds before a connection will be considered timed out
+     */
+    public int connectTimeoutMs() {
+        return connectTimeoutMs;
+    }
+
+    /**
+     * @return number of seconds between TCP keepalive packets being sent to the peer
+     */
+    public int keepAliveIntervalSecs() {
+        return keepAliveIntervalSecs;
+    }
+
+    /**
+     * @return number of seconds to wait for a keepalive response before considering the connection timed out
+     */
+    public int keepAliveTimeoutSecs() {
+        return keepAliveTimeoutSecs;
+    }
+
+    public static Builder builder() {
+        return new DefaultSocketOptionsConfigurationBuilder();
+    }
+
+    /**
+     * A builder for {@link SocketOptionsConfiguration}.
+     *
+     * <p>All implementations of this interface are mutable and not thread safe.</p>
+     */
+    public interface Builder {
+
+        /**
+         * Sets the socket domain
+         * @param domain socket domain
+         * @return Builder
+         */
+        Builder domain(SocketOptions.SocketDomain domain);
```

Comment:
Do we need `domain` and `type`? If I understand correctly, there's no use-case for configuring these two with AWS services, right? We can always add them in the future if needed.

Reply:
👍

Reply:
I ended up keeping them for the advanced users, since it's all going to the same place in the CRT, but decoupled from both connection timeouts, and the keep-alive configuration so that SDK users discovering this can get their iteratively

Reply:
IMO we should only expose public APIs if needed to keep it a 2-way door because it'd be a lot harder to remove/change them in the future.

Reply:
Are you saying I should remove `SocketOptionsConfiguration` from this change so we can decide later how to expose them? I'm fine with that